app/src/org/commcare/android/tasks/UpdateTask.java

```diff
@@ -0,0 +1,266 @@
+                    singletonRunningInstance.getStatus() == Status.RUNNING) {
+                return singletonRunningInstance;
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Attaches pinned notification with a progress bar the task, which will
+     * report updates to and close down the notification.
+     *
+     * @param ctx For launching notification and localizing text.
+     */
+    public void startPinnedNotification(Context ctx) {
+        this.ctx = ctx;
+
+        pinnedNotificationProgress =
+                new PinnedNotificationWithProgress(ctx, "updates.pinned.download",
+                        "updates.pinned.progress", R.drawable.update_download_icon);
+
+    }
+
+    @Override
+    protected final AppInstallStatus doInBackground(String... params) {
+        profileRef = params[0];
+
+        setupUpdate();
+
+        try {
+            return stageUpdate();
+        } catch (Exception e) {
+            ResourceInstallUtils.logInstallError(e,
+                    "Unknown error ocurred during install|");
+            return AppInstallStatus.UnknownFailure;
+        }
+    }
+
+    private void setupUpdate() {
+        ResourceInstallUtils.recordUpdateAttemptTime(app);
+        resourceManager.incrementUpdateAttempts();
+
+        app.setupSandbox();
+
+        Logger.log(AndroidLogger.TYPE_RESOURCES,
+                "Beginning install attempt for profile " + profileRef);
+    }
+
+    private AppInstallStatus stageUpdate() {
+        Resource profile = resourceManager.getMasterProfile();
+        boolean appInstalled = (profile != null &&
+                profile.getStatus() == Resource.RESOURCE_STATUS_INSTALLED);
+
+        if (!appInstalled) {
+            return AppInstallStatus.UnknownFailure;
+        }
+
+        String profileRefWithParams =
+                ResourceInstallUtils.addParamsToProfileReference(profileRef);
+
+        return resourceManager.checkAndPrepareUpgradeResources(profileRefWithParams);
+    }
+
+    @Override
+    protected void onProgressUpdate(Integer... values) {
+        super.onProgressUpdate(values);
+
+        if (pinnedNotificationProgress != null) {
+            pinnedNotificationProgress.handleTaskUpdate(values);
+        }
+        if (taskListener != null) {
+            taskListener.handleTaskUpdate(values);
+        }
+    }
+
+    @Override
+    protected void onPostExecute(AppInstallStatus result) {
+        super.onPostExecute(result);
+
+        boolean inIncompleteState = !(result == AppInstallStatus.UpdateStaged ||
+                result == AppInstallStatus.UpToDate);
+
+        if (inIncompleteState) {
+            resourceManager.processUpdateFailure(result, ctx);
+        }
+
+        if (pinnedNotificationProgress != null) {
+            pinnedNotificationProgress.handleTaskCompletion(result);
+        }
+        if (taskListener != null) {
+            taskListener.handleTaskCompletion(result);
+        }
+
+        synchronized (lock) {
```

Comment:
Am wondering whether it makes sense for this lock to actually start at the beginning of this method, since it's not like something can usefully jump back onto this Task once the onPostExecute phase has begun?